internal/controllers/nmc_reconciler.go

```diff
@@ -555,104 +565,132 @@ func (h *nmcReconcilerHelperImpl) SyncStatus(ctx context.Context, nmcObj *kmmv1b
-		_, inStatus := statusLabels[nsn]
-		if !inSpec && !inStatus {
-			meta.RemoveLabel(
-				&node,
-				utils.GetKernelModuleReadyNodeLabel(nsn.Namespace, nsn.Name),
-			)
+	nodeModuleReadyLabels = h.lph.GetKernelModuleReadyLabels(nodeModuleReadyLabels, *node)
+	deprecatedNodeModuleReadyLabels = h.lph.GetDeprecatedKernelModuleReadyLabels(deprecatedNodeModuleReadyLabels, *node)
 
-			unloaded = append(unloaded, nsn)
-		}
-	}
+	// get spec labels and their config
+	specLabels := make(map[types.NamespacedName]kmmv1beta1.ModuleConfig)
+	specLabels = h.lph.GetSpecLabelsAndTheirConfigs(specLabels, nmc)
 
-	// v1 ready labels, deprecated - should be removed
-	for label := range deprecatedNodeModuleReadyLabels {
-		meta.RemoveLabel(&node, label)
-	}
+	// get status labels and their config
+	statusLabels := make(map[types.NamespacedName]kmmv1beta1.ModuleConfig)
+	statusLabels = h.lph.GetStatusLabelsAndTheirConfigs(statusLabels, nmc)
+
+	unloaded := make([]types.NamespacedName, 0, len(nodeModuleReadyLabels))
+	loaded := make([]types.NamespacedName, 0, len(specLabels))
+
+	// label in node but not in spec or status - should be removed
+	unloaded = h.removeOrphanedLabels(nodeModuleReadyLabels, specLabels, statusLabels, unloaded)
 
 	// label in spec and status and config equal - should be added
-	for nsn, specConfig := range specLabels {
-		statusConfig, ok := statusLabels[nsn]
-		if ok && reflect.DeepEqual(specConfig, statusConfig) && !nodeModuleReadyLabels.Has(nsn) {
-			meta.SetLabel(
-				&node,
-				utils.GetKernelModuleReadyNodeLabel(nsn.Namespace, nsn.Name),
-				"",
-			)
+	loaded = h.addEqualLabels(nodeModuleReadyLabels, specLabels, statusLabels, loaded)
 
-			loaded = append(loaded, nsn)
-		}
+	if err := h.nodeAPI.UpdateNodeLabels(ctx, deprecatedNodeModuleReadyLabels, loaded, unloaded, node); err != nil {
+		return nil, nil, fmt.Errorf("could not patch node: %v", err)
 	}
 
-	if err := h.client.Patch(ctx, &node, patchFrom); err != nil {
-		return fmt.Errorf("could not patch node: %v", err)
-	}
+	return loaded, unloaded, nil
+}
 
-	for _, nsn := range unloaded {
+func (h *nmcReconcilerHelperImpl) RecordEvents(loaded []types.NamespacedName, unloaded []types.NamespacedName, node *v1.Node) {
+	for _, nsn := range loaded {
 		h.recorder.AnnotatedEventf(
-			&node,
+			node,
 			map[string]string{"module": nsn.String()},
 			v1.EventTypeNormal,
-			"ModuleUnloaded",
-			"Module %s unloaded from the kernel",
+			"ModuleLoaded",
+			"Module %s loaded into the kernel",
 			nsn.String(),
 		)
 	}
-
-	for _, nsn := range loaded {
+	for _, nsn := range unloaded {
 		h.recorder.AnnotatedEventf(
-			&node,
+			node,
 			map[string]string{"module": nsn.String()},
 			v1.EventTypeNormal,
-			"ModuleLoaded",
-			"Module %s loaded into the kernel",
+			"ModuleUnloaded",
+			"Module %s unloaded from the kernel",
 			nsn.String(),
 		)
 	}
+}
 
-	return nil
+func (h *nmcReconcilerHelperImpl) removeOrphanedLabels(nodeModuleReadyLabels sets.Set[types.NamespacedName],
```

Comment:
Any reason this method isn't part of the help interface as well?

Reply:
I thought because its an helper (inner) function, so if someone wants to implement this interface outside of this package, we wont have to force him to implement inner functions

Reply:
You main point is correct, but we can say it about all the methods in the helper function. No one else will implement that interface and we only do it for easier unit-testing so having those methods in the interface is a better option IMO.